### PR TITLE
OCM-5281 | Feat | Add region validation from ocm-shards and list regions command

### PR DESCRIPTION
**Pull request by tirthct**
**Changed**
- Modified ocm login command to pass the region name (with or without rh prefix)
- Add package to get and validate regions
- Add list regions command

**Tested**
- Region name passed in should be validated from ocm-shards
- list rh-regions should show all available region names

### CODE REVIEW

File: cmd/ocm/login/cmd.go [hunk 1]
```diff
@@ -265,9 +275,14 @@ func run(cmd *cobra.Command, argv []string) error {
 		cfg.User = ""
 		cfg.Password = ""
 	}
-	err = config.Save(cfg)
-	if err != nil {
-		return fmt.Errorf("Can't save config file: %v", err)
+
+	// If an OCM region is provided, update the config URL with the SDK generated URL
+	if cfg.RhRegion != "" {
+		regValue, err := rhRegion.GetRhRegion(context.Background(), refreshToken, args.rhRegion)
```
Comment (macgregor) [inline cmd/ocm/login/cmd.go:281]: so i think region discovery inside login creates a awkward scenario if you try to use `--rh-region` without an active login session. So like say i have never logged in ever. I run `ocm login --rh-region=singapore` and im met with an error "Not logged in, run the 'login' command" well thats what im trying to do...

ocm-shards.json is public, we dont actually need credentials to fetch it, but we do need a url to fetch it from (e.g. api.openshift.com/.../ocm-shards.json for prod or api.stage.openshift.com/.../ocm-shards.json for stage, etc). 

I wonder how we could make it friendlier...

File: cmd/ocm/list/cmd.go [hunk 2]
```diff
@@ -46,4 +47,5 @@ func init() {
 	Cmd.AddCommand(upgradepolicy.Cmd)
 	Cmd.AddCommand(user.Cmd)
 	Cmd.AddCommand(version.Cmd)
+	Cmd.AddCommand(rhRegion.Cmd)
```
Comment (macgregor) [inline cmd/ocm/list/cmd.go:50]: is there a way to make the command hidden like the `--rh-region` flag?

File: pkg/config/config.go [hunk 3]
```diff
@@ -48,6 +48,7 @@ type Config struct {
 	RefreshToken string   `json:"refresh_token,omitempty" doc:"Offline or refresh token."`
 	Scopes       []string `json:"scopes,omitempty" doc:"OpenID scope. If this option is used it will replace completely the default scopes. Can be repeated multiple times to specify multiple scopes."`
 	TokenURL     string   `json:"token_url,omitempty" doc:"OpenID token URL."`
+	RhRegion     string   `json:"rh_region,omitempty" doc:"OCM region name."`
```
Comment (macgregor) [inline pkg/config/config.go:51]: i dont think we need to store this in the config since its implicit in the url. `ocm login --rh-region=singapore` results in the "singapore" region ocm instance url being stored in URL config variable. We want to try to limit region awareness to just the login command for now until we find a good need to explicitly store the selected region for use in other parts of the cli.

File: pkg/rhRegion/rhRegion.go [hunk 4]
```diff
@@ -0,0 +1,50 @@
+package rhRegion
```
Comment (macgregor) [inline pkg/rhRegion/rhRegion.go:1]: so this whole file is a great example of what can get pushed down into the ocm-sdk for reuse in rosa cli. not necessarily now, but keep it in mind.

File: pkg/rhRegion/rhRegion.go [hunk 5]
```diff
@@ -0,0 +1,50 @@
...
+
+type Region struct {
+	URL string
+	AWS []string
+	GCP []string
+}
+
+func GetRhRegions(ctx context.Context, token string) (map[string]Region, error) {
+	var regions map[string]Region
+	globalConn, err := sdk.NewConnectionBuilder().Tokens(token).RhRegion("integration").BuildContext(ctx)
```
Comment (macgregor) [inline pkg/rhRegion/rhRegion.go:20]: couple things:
1. ocm-shards.json is public, its doesnt actually need a token, you just need a url
2. forget about RhRegion on the connection builder, it should probably be removed, it doesnt make sense there

So see if you can make this function signature work `func GetRhRegions(ctx context.Context, URL String) (map[string]Region, error)`. And then later we can try to refactor `GetRhRegions` and `GetRhRegion` down into the ocm-sdk itself.

File: pkg/rhRegion/rhRegion.go [hunk 6]
```diff
@@ -0,0 +1,50 @@
...
+
+	return regions, nil
+}
+
+func GetRhRegion(ctx context.Context, token string, regionName string) (Region, error) {
+	regions, err := GetRhRegions(ctx, token)
+	if err != nil {
+		return Region{}, err
+	}
+	if !strings.HasPrefix(regionName, "rh-") {
```
Comment (macgregor) [inline pkg/rhRegion/rhRegion.go:42]: I understand the impulse to be flexible or helpful, but in general try to avoid "smart keys" where you support multiple key variations with application logic, it ends up being a pain in the butt to maintain long term and is generally unintuitive to users. 

Speaking from experience :)

File: cmd/ocm/login/cmd.go (hunk 1 above)
Comment (tirthct) [reply cmd/ocm/login/cmd.go:281]: If we pass ocm login --rh-region=singapore, I get this message currently:
```
In order to log in it is mandatory to use '--token', '--user' and '--password', or '--client-id' and '--client-secret'.
You can obtain a token at: https://console.redhat.com/openshift/token .
See 'ocm login --help' for full help.
```

Also, token is mandatory right now if I want to create an SDK connection using ConnectionBuilder i.e. 

> globalConn, err := sdk.NewConnectionBuilder().Tokens(token).RhRegion("integration").BuildContext(ctx)

If I try using this command without the token, I get an error 
`either a token, an user name and password or a client identifier and secret are necessary, but none has been provided`

We can use the direct URLs (i.e. instead of sdk just call the api.openshift.com/.../ocm-shards.json directly, but since this might move to SDK, not sure if this wold be a good idea or not

File: cmd/ocm/list/cmd.go (hunk 2 above)
Comment (tirthct) [reply cmd/ocm/list/cmd.go]: Hid the command

File: pkg/config/config.go (hunk 3 above)
Comment (tirthct) [reply pkg/config/config.go]: Removed the JSON representation. Should we omit this?

File: pkg/rhRegion/rhRegion.go (hunk 4 above)
Comment (tirthct) [reply pkg/rhRegion/rhRegion.go:1]: Ack.

File: pkg/rhRegion/rhRegion.go (hunk 6 above)
Comment (tirthct) [reply pkg/rhRegion/rhRegion.go]: Removed the rh- prefix check

File: pkg/rhRegion/rhRegion.go (hunk 5 above)
Comment (tirthct) [reply pkg/rhRegion/rhRegion.go]: Changed sdk call to simple http get method

File: pkg/rhRegion/rhRegion.go [hunk 7]
```diff
@@ -0,0 +1,60 @@
...
+	if err != nil {
+		return "", err
+	}
+	regionDiscoveryHost := "api.openshift.com"
+	if strings.HasSuffix(baseUrl.Hostname(), "integration.openshift.com") {
+		regionDiscoveryHost = "api.integration.openshift.com"
+	} else if strings.HasSuffix(baseUrl.Hostname(), "integration.openshift.com") {
+		regionDiscoveryHost = "api.stage.openshift.com"
+	}
+	regionDiscoveryHost = "api.integration.openshift.com"
```
Comment (macgregor) [inline pkg/rhRegion/rhRegion.go:58]: this temporary while we play in integration? add a `//TODO: remove before merge` if so

Comment (tirthct) [reply pkg/rhRegion/rhRegion.go]: Added TODO

File: cmd/ocm/login/cmd.go [hunk 8]
```diff
@@ -239,6 +247,7 @@ func run(cmd *cobra.Command, argv []string) error {
 
 	// Update the configuration with the values given in the command line:
 	cfg.TokenURL = tokenURL
+	cfg.RhRegion = args.rhRegion
```
Comment (macgregor) [inline cmd/ocm/login/cmd.go:250]: again, no need to persist the `RhRegion` to the config. We go from args.rhRegion -> region discover -> cfg.URL, once we have the URL we are done with region. If we find a specific need later where we need to keep track of the name of the region for some purpose, we can add it to the persisted config then. For now, I dont see it being needed after line 284 (`cfg.URL = regValue.URL`)

File: pkg/config/config.go [hunk 9]
```diff
@@ -48,9 +48,10 @@ type Config struct {
 	RefreshToken string   `json:"refresh_token,omitempty" doc:"Offline or refresh token."`
 	Scopes       []string `json:"scopes,omitempty" doc:"OpenID scope. If this option is used it will replace completely the default scopes. Can be repeated multiple times to specify multiple scopes."`
 	TokenURL     string   `json:"token_url,omitempty" doc:"OpenID token URL."`
-	URL          string   `json:"url,omitempty" doc:"URL of the API gateway. The value can be the complete URL or an alias. The valid aliases are 'production', 'staging' and 'integration'."`
-	User         string   `json:"user,omitempty" doc:"User name."`
-	Pager        string   `json:"pager,omitempty" doc:"Pager command, for example 'less'. If empty no pager will be used."`
+	RhRegion     string
```
Comment (macgregor) [inline pkg/config/config.go:51]: kill it

File: pkg/config/config.go [hunk 10]
```diff
@@ -270,6 +271,10 @@ func (c *Config) Connection() (connection *sdk.Connection, err error) {
 	}
 	builder.Insecure(c.Insecure)
 
+	if c.RhRegion != "" {
+		builder.RhRegion(c.RhRegion)
```
Comment (macgregor) [inline pkg/config/config.go:275]: kill it. We connect to the correct region because cfg.URL is pointed to the correct region url. The sdk ConnectionBuilder will not be aware of the concept of a region.

File: pkg/rhRegion/rhRegion.go [hunk 11]
```diff
@@ -0,0 +1,61 @@
...
+func DetermineRegionDiscoveryUrl(ocmServiceUrl string) (string, error) {
+	baseUrl, err := url.Parse(ocmServiceUrl)
+	if err != nil {
+		return "", err
+	}
+	regionDiscoveryHost := "api.openshift.com"
+	//TODO: Remove the OR condition from this if statement before the MR merge
+	if strings.HasSuffix(baseUrl.Hostname(), "integration.openshift.com") || true {
+		regionDiscoveryHost = "api.integration.openshift.com"
+	} else if strings.HasSuffix(baseUrl.Hostname(), "integration.openshift.com") {
```
Comment (macgregor) [inline pkg/rhRegion/rhRegion.go:57]: stage.openshift.com

File: cmd/ocm/login/cmd.go [hunk 12]
```diff
@@ -265,6 +274,16 @@ func run(cmd *cobra.Command, argv []string) error {
 		cfg.User = ""
 		cfg.Password = ""
 	}
+
+	// If an OCM region is provided, update the config URL with the SDK generated URL
+	if cfg.RhRegion != "" {
+		regValue, err := rhRegion.GetRhRegion(connection.URL(), args.rhRegion)
+		if err != nil {
+			return fmt.Errorf("Can't find region: %v", err)
```
Comment (macgregor) [inline cmd/ocm/login/cmd.go:282]: nit: fyi, you should wrap errors with `%w` rather than `%v`. `%w` provides more functionality and preserves stack traces better. people do it incorrectly all over our code base so i point it out when i see it: https://go.dev/blog/go1.13-errors

File: cmd/ocm/list/rhRegion/cmd.go [hunk 13]
```diff
@@ -0,0 +1,53 @@
...
+	}
+
+	connection, err := ocm.NewConnection().Build()
+	if err != nil {
+		return err
+	}
+	defer connection.Close()
+
+	// Check that the configuration has credentials or tokens that don't have expired:
+	armed, reason, err := cfg.Armed()
```
Comment (macgregor) [inline cmd/ocm/list/rhRegion/cmd.go:37]: i dont think we need the armed check or an ocm-sdk ConnectionBuilder since ocm-shards is public (anonymous). so i think we can boil this command down to just:
```

func run(cmd *cobra.Command, argv []string) error {
	// Load the configuration file:
	cfg, err := config.Load()
	if err != nil {
		return fmt.Errorf("Can't load config file: %v", err)
	}
	if cfg == nil {
		return fmt.Errorf("Not logged in, run the 'login' command")
	}

	regions, err := rhRegion.GetRhRegions(cfg.URL)
	if err != nil {
		return fmt.Errorf("Failed to get OCM regions: %v", err)
	}

	for regionName := range regions {
		fmt.Println(regionName)
	}
	return nil
```

File: cmd/ocm/login/cmd.go (hunk 8 above)
Comment (tirthct) [reply cmd/ocm/login/cmd.go]: Ack. Removed

File: go.mod [hunk 14]
```diff
@@ -72,9 +73,10 @@ require (
...
 	github.com/zgalor/weberr v0.7.0 // indirect
 	golang.org/x/crypto v0.1.0 // indirect
-	golang.org/x/net v0.10.0 // indirect
 	golang.org/x/sys v0.10.0 // indirect
 	golang.org/x/tools v0.9.3 // indirect
 	google.golang.org/protobuf v1.28.1 // indirect
 	gopkg.in/yaml.v2 v2.4.0 // indirect
 )
+
+replace github.com/openshift-online/ocm-sdk-go v0.1.388 => ../ocm-sdk-go
```
[the remark this answers is not in the document]
Comment (tirthct) [reply go.mod]: Removed since we no longer need RhRegion

File: pkg/config/config.go (hunk 9 above)
Comment (tirthct) [reply pkg/config/config.go]: Removed

File: pkg/config/config.go (hunk 10 above)
Comment (tirthct) [reply pkg/config/config.go]: Removed

File: pkg/rhRegion/rhRegion.go (hunk 11 above)
Comment (tirthct) [reply pkg/rhRegion/rhRegion.go]: Fixed

File: cmd/ocm/login/cmd.go (hunk 12 above)
Comment (tirthct) [reply cmd/ocm/login/cmd.go]: Ahh, I see! Changed to %w

File: cmd/ocm/list/rhRegion/cmd.go (hunk 13 above)
Comment (tirthct) [reply cmd/ocm/list/rhRegion/cmd.go]: Ack. Removed the check

**Comment by macgregor** [comment]
lgtm, thanks Tirth!

**Comment by macgregor** [comment]
lgtm

File: cmd/ocm/list/cmd.go [hunk 15]
```diff
@@ -46,4 +47,7 @@ func init() {
 	Cmd.AddCommand(upgradepolicy.Cmd)
 	Cmd.AddCommand(user.Cmd)
 	Cmd.AddCommand(version.Cmd)
+	rhRegionCmd := rhRegion.Cmd
+	rhRegionCmd.Hidden = true
```
Comment (markturansky) [inline cmd/ocm/list/cmd.go:51]: should this be set by default in the cmd?  seems odd to toggle off here.

Comment (tirthct) [reply cmd/ocm/list/cmd.go:51]: We don't want the command `ocm list rh-regions` to be visible. We can hide the flags but this is the only way to hide the subcommands : https://github.com/spf13/cobra/issues/1507#issuecomment-1632443187

Comment (markturansky) [reply cmd/ocm/list/cmd.go:51]: if this is always hidden by default, i'm suggesting you default to hidden/true in the cmd itself.  it seems weird to toggle at this level.

Comment (tirthct) [reply cmd/ocm/list/cmd.go]: Ack. Added hidden in Cmd itself

**Review by markturansky — APPROVED** [review]
LGTM

**Review by markturansky — APPROVED** [review]
LGTM

File: cmd/ocm/list/rhRegion/cmd.go [hunk 16]
```diff
@@ -0,0 +1,40 @@
...
+
+	"github.com/openshift-online/ocm-cli/pkg/config"
+	"github.com/spf13/cobra"
+)
+
+var Cmd = &cobra.Command{
+	Use:     "rh-regions",
+	Short:   "List available OCM regions",
+	Long:    "List available OCM regions",
+	Example: `  # List all supported OCM regions ocm list rh-regions`,
```
Comment (cristianoveiga) [inline cmd/ocm/list/rhRegion/cmd.go:15]: should this be `ocm list rh-regions` only?

File: go.mod [hunk 17]
```diff
@@ -1,6 +1,6 @@
 module github.com/openshift-online/ocm-cli
 
-go 1.18
+go 1.21
```
Comment (cristianoveiga) [inline go.mod:3]: is the go version update required here? I'm wondering if it can/should be handled in a separate MR.

Comment (tirthct) [reply go.mod:3]: OCM SDK uses latest golang : https://github.com/openshift-online/ocm-sdk-go/commit/1252182841bbe28fc65153d4c502d95a67b158cf#diff-33ef32bf6c23acb95f5902d7097b7a1d5128ca061167ec0716715b0b9eeaa5f6
We can create a separate MR here though.

File: cmd/ocm/list/rhRegion/cmd.go (hunk 16 above)
Comment (tirthct) [reply cmd/ocm/list/rhRegion/cmd.go]: Example sections in other files have the same format : https://github.com/openshift-online/ocm-cli/blob/main/cmd/ocm/list/version/cmd.go#L38

Comment (cristianoveiga) [reply cmd/ocm/list/rhRegion/cmd.go]: Okay, but IMO that's not really helpful.

**Review by cristianoveiga — APPROVED** [review]
LGTM
